sycl/include/sycl/ext/oneapi/matrix/matrix-unified.hpp

```diff
@@ -43,73 +43,24 @@ struct joint_matrix {
   }
 };
 
-#ifdef __SYCL_DEVICE_ONLY__
-template <typename Group, typename T, use Use, size_t Rows, size_t Cols,
-          layout Layout>
-class wi_data {
-
-  joint_matrix<Group, T, Use, Rows, Cols, Layout> &jm;
-
-  wi_data(joint_matrix<Group, T, Use, Rows, Cols, Layout> &_jm) : jm(_jm){};
-
-  template <typename Grp, typename Type, use UseJm, size_t NumRows,
-            size_t NumCols, layout LayoutJm>
-  friend decltype(auto)
-  get_wi_data(Grp,
-              joint_matrix<Grp, Type, UseJm, NumRows, NumCols, LayoutJm> &);
-
-public:
-  size_t length() {
-#if defined(__NVPTX__)
-    return jm.cuda_impl.wi_marray.size();
-#else
-    return __spirv_JointMatrixWorkItemLengthINTEL(jm.spvm);
-#endif
-  };
-
-  decltype(auto) operator[](size_t i) {
+template <typename Group, typename T, use Use, size_t M, size_t N,
+          layout Layout, typename F>
+inline __SYCL_ALWAYS_INLINE void
+joint_matrix_apply(Group sg, joint_matrix<Group, T, Use, M, N, Layout> &C,
+                   F &&lambda) {
+#if defined(__SYCL_DEVICE_ONLY__)
 #if defined(__NVPTX__)
```

Comment:
This is the cuda implementation. Happy for you to add it here.
```suggestion
#if defined(__NVPTX__)
  for (int i = 0; i < jm.cuda_impl.wi_marray.size(); i++) {
    lambda(jm.cuda_impl.wi_marray[i]);
  }
```